ftw/publisher/core/adapters/ftw_shop.py

```diff
@@ -122,3 +123,69 @@ def setData(self, data, metadata):
             variations_config.purge_dict()
         else:
             variations_config.updateVariationConfig(data)
+
+
+class ShopCategorizableReferences(object):
+    """DataCollector adapter for the category references of the categorizable shop types.
+    """
+
+    implements(IDataCollector)
+    logger = getLogger()
+    security = ClassSecurityInformation()
+
+    def __init__(self, obj):
+        self.object = obj
+
+    security.declarePrivate('getData')
+    def getData(self):
+        categories_uids = [reference.UID() for reference in self.object.listCategories()]
+        data = make_serializable(categories_uids)
+        return data
+
+    security.declarePrivate('setData')
+    def setData(self, data, metadata):
+        self.logger.info('Updating shop item category references (UID %s)' %
+                         (self.object.UID()))
+
+        uid_catalog = getToolByName(self.object, 'uid_catalog')
+
+        # Add categories available on source.
+        for category_uid in data:
+            brains = uid_catalog(UID=category_uid)
+            if brains:
+                self.object.addToCategory(brains[0].getObject())
+
+        # Remove categories no longer available on the source.
+        existing_categories = [reference.UID() for reference in self.object.listCategories()]
+        categories_to_be_removed = set(existing_categories) - set(data)
+        for category in categories_to_be_removed:
+            self.object.removeFromCategory(category)
+
+
+class ShopCategorizableRanks(object):
+    """DataCollector adapter for the (category) ranks of the categorizable shop types.
+    """
+
+    implements(IDataCollector)
+    logger = getLogger()
+    security = ClassSecurityInformation()
+
+    def __init__(self, obj):
+        self.object = obj
+
+    security.declarePrivate('getData')
+    def getData(self):
+        if hasattr(self.object, '_categoryRanks'):
+            data = deepcopy(self.object._categoryRanks)
```

Comment:
Why is the `deepcopy` necessary? As far as I understand the code, data is not mutated.